### PR TITLE
Return non-canonical block headers from finalized blocks.

### LINKED ISSUE #5127: GetBlockHeaders should return non canonical block headers if appropriate
GetBlockHeaders needs to use a selector that allows non canonical blocks so that it can respond with canonical=false in the case when block headers are not on the canonical chain.

This will also mean multiple objects can be output, so it'll require a little rework.

Follows on from #5080 where non canonical blocks may come out of GetBlockHeader.

### PR DESCRIPTION
**Pull request by rolfyone**
fixes #5127

Signed-off-by: Paul Harris <paul.harris@consensys.net>

## Documentation

- [X] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [X] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: data/provider/src/main/java/tech/pegasys/teku/api/blockselector/BlockSelectorFactory.java [hunk 1]
```diff
@@ -95,12 +95,8 @@ public BlockSelector nonCanonicalBlocksSelector(final UInt64 slot) {
...
       final ChainHead chainHead = maybeChainHead.get();
       return client
-          .getAllBlocksAtSlot(slot)
-          .thenApply(
-              blocks ->
-                  blocks.stream()
-                      .map(block -> lookupBlockData(block, chainHead))
-                      .collect(Collectors.toList()));
+          .getAllBlocksAtSlot(slot, chainHead.getRoot())
+          .thenApply(blocks -> new ArrayList<>(blocks));
```
Comment (ajsutton) [inline data/provider/src/main/java/tech/pegasys/teku/api/blockselector/BlockSelectorFactory.java]: Given `getAllBlocksAtSlot` is only used here we probably should modify it to just return a `List` directly and avoid the need to copy.

File: storage/src/main/java/tech/pegasys/teku/storage/client/CombinedChainDataClient.java [hunk 2]
```diff
@@ -535,22 +536,53 @@ public boolean isOptimisticBlock(final Bytes32 blockRoot) {
...
+    canonicalBlock.ifPresent(block -> result.add(toBlockAndMetaData(block, true)));
+    return result;
+  }
+
+  private BlockAndMetaData toBlockAndMetaData(
+      final SignedBeaconBlock signedBeaconBlock, final boolean canonical) {
+    return new BlockAndMetaData(
+        signedBeaconBlock,
+        spec.atSlot(signedBeaconBlock.getSlot()).getMilestone(),
+        isOptimisticBlock(signedBeaconBlock.getRoot()),
```
Comment (ajsutton) [inline storage/src/main/java/tech/pegasys/teku/storage/client/CombinedChainDataClient.java]: This isn't quite right.  We should be marking blocks as optimistic if either the block is optimistic or the chain head is optimistic.

**Review by ajsutton — APPROVED** [review]
LGTM.

File: data/provider/src/main/java/tech/pegasys/teku/api/blockselector/BlockSelectorFactory.java (hunk 1 above)
Comment (rolfyone) [reply data/provider/src/main/java/tech/pegasys/teku/api/blockselector/BlockSelectorFactory.java]: this one we'll put into another PR just to be clear on this comment

**Pull request merged**